crypto/src/main/scala/org/bitcoins/crypto/MuSig2Util.scala

```diff
@@ -0,0 +1,225 @@
+package org.bitcoins.crypto
+
+import scodec.bits.ByteVector
+
+// TODO partial signature verification
+// TODO implement tweaking
+object MuSig2Util {
+
+  private val nonceNum: Int = 2
+
+  type MultiNoncePriv = Vector[ECPrivateKey]
+  type MultiNoncePub = Vector[ECPublicKey]
+
+  case class KeyGenContext(
+      aggPubKey: ECPublicKey,
+      tacc: FieldElement,
+      gacc: Boolean) {
+    def getPubKey: ByteVector = aggPubKey.bytes.tail
+    def getPubKey33: ByteVector = aggPubKey.bytes
+  }
+
+  case class KeySet(keys: Vector[SchnorrPublicKey]) {
+    keys.init.zip(keys.tail).foreach { case (key1, key2) =>
+      require(key1.hex.compareTo(key2.hex) <= 0,
+              "Keys must be sorted lexicographically")
+    }
+
+    lazy val serialize: ByteVector = {
+      keys.map(_.bytes).reduce(_ ++ _)
+    }
+
+    def keyAggCoef(key: SchnorrPublicKey): FieldElement = {
+      if (key == secondKey) FieldElement.one
+      else {
+        val listHashBytes = aggListHash(serialize)
+        val bytes = aggCoefHash(listHashBytes ++ key.bytes)
+
+        FieldElement(new java.math.BigInteger(1, bytes.toArray))
+      }
+    }
+
+    lazy val aggPubKey: ECPublicKey = {
+      keys
+        .map { key =>
+          val coef = keyAggCoef(key)
+          key.publicKey.multiply(coef)
+        }
+        .reduce(_.add(_))
+    }
+
+    lazy val secondKey: SchnorrPublicKey = keys.tail.head
+  }
+
+  object KeySet {
+
+    def apply(keys: Vector[SchnorrPublicKey]): KeySet = {
+      val sortedKeys = keys.sorted(Ordering.by[SchnorrPublicKey, String](_.hex))
+      new KeySet(sortedKeys)
+    }
+
+    def apply(keys: SchnorrPublicKey*): KeySet = {
+      KeySet(keys.toVector)
+    }
+  }
+
+  def aggListHash(bytes: ByteVector): ByteVector = {
+    CryptoUtil.taggedSha256(bytes, "MuSig2/agg_list").bytes
+  }
+
+  def aggCoefHash(bytes: ByteVector): ByteVector = {
+    CryptoUtil.taggedSha256(bytes, "MuSig2/agg_coef").bytes
+  }
+
+  def nonHash(bytes: ByteVector): ByteVector = {
+    CryptoUtil.taggedSha256(bytes, "MuSig2/non").bytes
+  }
+
+  def nonCoefHash(bytes: ByteVector): ByteVector = {
+    CryptoUtil.taggedSha256(bytes, "MuSig2/non_coef").bytes
+  }
+
+  def sigHash(bytes: ByteVector): ByteVector = {
+    CryptoUtil.taggedSha256(bytes, "MuSig2/sig").bytes
+  }
+
+  def auxHash(bytes: ByteVector): ByteVector = {
+    CryptoUtil.taggedSha256(bytes, "MuSig2/aux").bytes
+  }
+
+  def genMultiNonce(
+      privKeyOpt: Option[ECPrivateKey] = None,
+      aggPubKeyOpt: Option[SchnorrPublicKey] = None,
+      msgOpt: Option[ByteVector] = None,
+      extraInOpt: Option[ByteVector] = None): (
+      MultiNoncePub,
+      MultiNoncePriv) = {
+    msgOpt.foreach(msg => require(msg.length == 32))
+    extraInOpt.foreach(extraIn => require(extraIn.length <= 4294967295L))
+
+    val preRand =
+      ECPrivateKey.freshPrivateKey.bytes // TODO this needs to draw 32 bytes uniformily at random
```

Comment:
is that not what this does?

Reply:
If I'm not mistaken it almost is, but this throws away a random draw if it is bigger than the field order (which is just slightly less than 2^32), so I need to rewrite without the throw away code

Reply:
Turns out we already have `CryptoUtil.randomBytes`, so I use this :)